lib/rjgit.rb

```diff
@@ -209,15 +210,38 @@ def self.grep(repository, query, options={})
 
       query = Regexp.new(query.source, query.options | Regexp::IGNORECASE) if case_insensitive
 
-      ls_tree(repo, nil, options.fetch(:ref, 'HEAD'), ls_tree_options).each_with_object({}) do |item, result|
-        blob = Blob.new(repo, item[:mode], item[:path], walk.lookup_blob(ObjectId.from_string(item[:id])))
-        next if blob.binary?
+      # We optimize below by first grepping the entire file, and then, if a match is found, then identifying the individual line.
+      # To avoid missing full-line matches during the optimization, we first convert multiline anchors to single-line anchors.
+      query = Regexp.new(query.source.gsub(/\A\\A/, '^').gsub(/\\z\z/, '$'), query.options)
 
-        rows = blob.data.split("\n")
+      ref = options.fetch(:ref, 'HEAD')
+      files_to_scan = ls_tree(repo, nil, ref, ls_tree_options)
+
+      files_to_scan.each_with_object({}) do |file, result|
+        id = if file[:mode] == SYMLINK_TYPE
+          symlink_source = repo.open(ObjectId.from_string(file[:id])).get_bytes.to_a.pack('c*').force_encoding('UTF-8')
+          unless symlink_source[File::SEPARATOR]
+            dir = file[:path].split(File::SEPARATOR)
+            dir[-1] = symlink_source
+            symlink_source = File.join(dir)
```

Comment:
What's happening here when you reassign `symlink_source`?

Reply:
Commit message says "Resolve source files of relative symlinks correctly".  From what I remember, it's because symlinks could be absolute or relative, but to open it we need an absolute path.  If there's no `/` in the path, that tells us it's relative, so we need to join it with the dirname to get the full path to it.

Reply:
Thanks for explaining. I guess I don't understand the reason for the intermediate `dir` array then. Why not instead do:
```ruby
symlink_source = ::File.join(file[:path], symlink_source) unless symlink_source.include?[::File::SEPARATOR]
```

Reply:
Thanks for explaining. I guess I don't understand the reason for the intermediate `dir` array then. Why not instead do:
```ruby
symlink_source = ::File.join(file[:path], symlink_source) unless symlink_source[::File::SEPARATOR]
```
(**EDIT** removed my `.include?` call for speed; your implementation of checking for '/' is faster.)

Reply:
`dir[-1]` (before I overwrite it) is the symlink file's name.  Given the following example:

```
$ ls -l /tmp/example
total 0
-rw-rw-r-- 1 my_user my_group 0 Nov 17 10:31 abc
lrwxrwxrwx 1 my_user my_group 3 Nov 17 10:31 def -> abc
```

`file[:path]` is '/tmp/example/def' and `symlink_source` is `abc`
`dir = file[:path].split(File::SEPARATOR)` is `['/tmp', 'example', 'def']`
`dir[-1] = symlink_source` changes `dir` to `['/tmp', 'example', 'abc']`
`symlink_source = File.join(dir)` results in '/tmp/example/abc'

Doing this in one line, as you propose, it would look like this:

```
symlink_source = file[:path].sub(/[^#{File::SEPARATOR}]+\z/, symlink_source) unless symlink_source[::File::SEPARATOR]
```

Personally, I find that less readable than the current implementation but I'll let you make the decision.

Reply:
Nope, I'm on board now! Thanks for helping me understand.